source/common/http/http2/codec_impl.cc

```diff
@@ -206,10 +206,15 @@ void ConnectionImpl::StreamImpl::resetStream(StreamResetReason reason) {
   if (local_end_stream_ && !local_end_stream_sent_) {
     parent_.pending_deferred_reset_ = true;
     deferred_reset_.value(reason);
+    ENVOY_CONN_LOG(trace, "deferred reset stream", parent_.connection_);
   } else {
     resetStreamWorker(reason);
-    parent_.sendPendingFrames();
   }
+
+  // We must still call sendPendingFrames() in both the deferred and not deferred path. This forces
+  // the cleanup logic to run which will reset the stream in all cases if all data frames could not
+  // be sent.
+  parent_.sendPendingFrames();
```

Comment:
Unfortunately we can't, because sendPendingFrames() will return immediately if we are currently dispatching read data (interleaving read/write is not possible w/ nghttp2). So we still need the flag for the currently dispatching case.